architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,641 @@
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+## User Experience (if applicable)
+
+See API design. Tooling and user-experiences are out of scope for this document.
+
+## Design
+
+### High Level Design
+
+The registration APIs allow a user to control the set of resource types known to UCP. Since UCP is responsible for routing, it needs to know the set of resource providers and resource types so that it can send them traffic. 
+
+While resource types are the primary concept for **consumers** of extensibility (eg: application developers using Radius), resource providers are a significant concept for users that **implement** extensibility. The API design proposed here includes resource providers as a concept, and supports important scenarios for operating and versioning resource providers over time, based on lessons-learned from ARM.
+
+For anyone skeptical about this, the concept of resource provider is already present in **every** resource id/URL. This is the meaning of the `providers` segment. Currently URLs like `/planes/radius/local/providers` don't have any functionality associated with them, and return a 404 - this proposal addresses this gap.
+
+One of these lessons-learned is the ability to register multiple regions/locations where a resource provider can operate. In particular users will need the ability to gradually roll out functionality across regions rather than updating all regions at once. This is complex, and doesn't help us in the short term. The approach taken here is aligned with our current strategy for region/location in Radius. The APIs are designed for multi-region but the implementation assumes a single region is used.
+
+#### Resource Types
+
+The registration APIs add a few new resource types:
+
+- `System.Resources/resourceProviders` for resource providers.
+- `System.Resources/resourceProviders/resourceTypes` for resource types.
+- `System.Resources/resourceProviders/resourceTypes/apiVersions` for API versions of resource types.
+- `System.Resources/resourceProviders/locations` for the configuration of a resource provider in a specific location/region.
+
+These new resource types define the full CRUDL set of APIs. These APIs are optimized for operations that need to modify or deeply understand resource types. 
+
+These are our first nested resource type in UCP/Radius. This is a good usecase for a nested resource types because:
+
+- The `apiVersions` body will be be large due to the inclusion of an OpenAPI schema. 
+- Users need granular registration and unregistration of individual resource types and API versions. (See scenario 2 above).
+- Modeling `location` as a resource optimizes for per-region rollouts and validation of new functionality.
+
+---
+
+These new resource types are scoped to the Radius plane. 
+
+> eg: `/planes/radius/local/providers/System.Resources/resourceProviders`
```

Comment:
I have a basic question here. what does `/local/` mean in the context of planes?

Reply:
Local is the sentinel value we use for "this is a single-tenant installation of Radius".